### ISSUE TITLE
[API Proposal]: Enumerable.ChunkBy

### ISSUE THREAD

**Issue by angellaa** [dotnet/runtime#79111]
### Background and motivation

The framework already contains a [Chunk](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Chunk.cs) extension method to splits the elements of a sequence into chunks of size at most size.

There are situations when you want to splits the elements of a sequence based on a predicate, similarly to using `String.Split` for strings.

I propose to introduce a ChunkBy extension method to chunk a sequence based on a predicate.

This is a naive reference implementation of the idea:

```csharp 

    public static IEnumerable<IEnumerable<TSource>> ChunkBy<TSource>(
        this IEnumerable<TSource> source, 
        Func<TSource, bool> predicate)
    {
        var chunk = new List<TSource>();

        foreach (var item in source)
        {
            if (predicate(item))
            {
                yield return chunk;
                chunk.Clear();
                continue;
            }

            chunk.Add(item);
        }

        yield return chunk;
    }

```

LINQ provides methods like [Max](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.max?view=net-7.0) and [MaxBy](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.maxby?view=net-7.0) so the name ChunkBy would follow the same patterns.

### API Proposal

```csharp
namespace System.Linq;

public static partial class Enumerable
{
    public static IEnumerable<IEnumerable<TSource>> Chunk<TSource>(this IEnumerable<TSource> source, Func<TSource,bool> predicate);
}
```

### API Usage

```csharp
        var sequence = new[] { "1", "2", "", "3", "4", "", "5" };

        foreach (var chunk in sequence.ChunkBy(x => x == ""))
        {
            Console.WriteLine(string.Join(',', chunk));
        }

        // Prints
        // 1,2
        // 3,4
        // 5
```

### Alternative Designs

To match the same API signature of Chunk, wecan use a TSource[] array to represent chunks in the return type.

```csharp
public static partial class Enumerable
{
    public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> source, Func<TSource,bool> predicate);
}
```

### Risks

This is a new extension method on Enumerable so I can't think of any risks of breaking changes.

**Comment by lsoft** [comment]
@angellaa we are loosing "border" items in this design, probably it's not a expected behaviour for `Chunk` operation (but I see the simmetry with `String.Split`). From my point of view, `ChunkBy` is a good idea and I see the place where I can use it for my codebase, but I'm pretty sure that I do not want to loose items in `ChunkBy`.

If we choose not to lost items, then the question arise: should we put the border item as the last element of the chunk, or a first element of the next chunk? I have no opinion.

**Comment by eiriktsarpalis** [comment]
The `By` suffix is typically reserved methods that are performing operations on the source enumerable based on project to a given key type. This takes a predicate and not project, so I would argue this might probably be named `GroupWhile`, folllowing existing convention of the `TakeWhile` method.

**Comment by angellaa** [comment]
I agree with you @lsoft, I think SplitBy feels a more appropiate name.

My intend was to add a function that behaves like Split for strings.

The predicate helps to identify what elements we consider separators.

Should I update the title and the content of the proposal?

**Comment by eiriktsarpalis** [comment]
Honestly, I'd be inclined to say that this doesn't meet the bar for inclusion. I had not realized that items satisfying the predicate are being skipped as "separators" -- I had assumed they were being included as the final element in the chunk. Not saying that it's "wrong" but it definitely feels like a helper method for a specific use case rather than something that is general-purpose enough.

**Comment by angellaa** [comment]
I agree that Chunk should not remove any elements from the sequence so reflecting back ChunkBy is the wrong name.

What I proposed was effectively Enumerable.Split but I totally accept that might not be general-purpose enough. 

I leave it with you to decide what to do. 

Pleasure to talk with you and keep up with the amazing work you're doing in making .NET better.

**Comment by theodorzoulias** [comment]
Two similar MoreLinq operators are the [`Segment`](https://github.com/morelinq/MoreLINQ/blob/master/MoreLinq/Segment.cs) and the [`Split`](https://github.com/morelinq/MoreLINQ/blob/master/MoreLinq/Split.cs). The first preserves
the separator, and the second consumes it.

```C#
var source = Enumerable.Range(1, 15);
var segmented = source.Segment(x => x % 5 == 0);
var split = source.Split(x => x % 5 == 0);
```

Output:

```none
Source:    1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15
Segmented: [1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15]
Split:     [1, 2, 3, 4], [6, 7, 8, 9], [11, 12, 13, 14]
```

[Live demo](https://dotnetfiddle.net/6sWdyb).

**Comment by angellaa** [comment]
@theodorzoulias the last example you shared `source.Split(x => x % 5 == 0)` covers effectively my use case.
